storage/repository/parse.go

```diff
@@ -0,0 +1,94 @@
+package repository
+
+import (
+	"net/url"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"fyne.io/fyne"
+)
+
+// NewFileURI implements the back-end logic to storage.NewFileURI, which you
+// should use instead. This is only here because other functions in repository
+// need to call it, and it prevents a circular import.
+//
+// Since: 2.0.0
+func NewFileURI(path string) fyne.URI {
+	// URIs are supposed to use forward slashes. On Windows, it
+	// should be OK to use the platform native filepath with UNIX
+	// or NT style paths, with / or \, but when we reconstruct
+	// the URI, we want to have / only.
+	if runtime.GOOS == "windows" {
+		// seems that sometimes we end up with
+		// double-backslashes
+		path = filepath.ToSlash(path)
+	}
+
+	return &uri{
+		scheme:        "file",
+		haveAuthority: true,
+		authority:     "",
+		path:          path,
+		query:         "",
+		fragment:      "",
+	}
+}
+
+// ParseURI implements the back-end logic for storage.ParseURI, which you
+// should use instead. This is only here because other functions in repository
+// need to call it, and it prevents a circular import.
+//
+// Since: 2.0.0
+func ParseURI(s string) (fyne.URI, error) {
+	// Extract the scheme.
+	scheme := ""
+	for i := 0; i < len(s); i++ {
+		if s[i] == ':' {
+			break
+		}
+		scheme += string(s[i])
+	}
+	scheme = strings.ToLower(scheme)
+
+	if scheme == "file" {
+		// Does this really deserve to be special? In principle, the
+		// purpose of this check is to pass it to NewFileURI, which
+		// allows platform path seps in the URI (against the RFC, but
+		// easier for people building URIs naively on Windows). Maybe
+		// we should punt this to whoever generated the URI in the
+		// first place?
+
+		path := s[5:] // everything after file:
+		if len(path) > 2 && path[:2] == "//" {
+			path = path[2:]
+		}
+
+		// Windows files can break authority checks, so just return the parsed file URI
+		return NewFileURI(path), nil
+	}
+
+	repo, err := ForURI(&uri{scheme: scheme})
```

Comment:
Perhaps this should be `ForScheme(string)` as that is specifically what the function cares about. Call sites with a URI become `ForScheme(uri.Scheme())`

Reply:
That sounds like exposing implementation detail through the API.
Isn't it possible that at some point in the future we may wish to expand on this, perhaps with the proposed combined repository (I can't recall the name).
I can't see the harm in staying with this API, as a scheme alone isn't something we've seen be used.

Reply:
I disagree on the basis that it would be extremely uncommon for someone to want a repository without having a URI they want to do something with.

Reply:
> That sounds like exposing implementation detail through the API.
> Isn't it possible that at some point in the future we may wish to expand on this, perhaps with the proposed combined repository (I can't recall the name).
> I can't see the harm in staying with this API, as a scheme alone isn't something we've seen be used.

This function is the complement of `func Register(scheme string, repository Repository)` which already exposes the "implementation detail" that repositories are registered to schemes, and if there were any future expansions of repositories mapped to more than just scheme then `Register` would need to be changed.

> I disagree on the basis that it would be extremely uncommon for someone to want a repository without having a URI they want to do something with.

"Someone" in this context would only be Fyne contributors as the docs explicitly discourage developers from calling this method directly. I would assume that this call site is the majority use case as `ParseURI` is the main entry point of this package, it doesn't have a URI and needs to allocate one just so it can wrap the scheme. All other call sites I can file have the URI and are therefore able to access the scheme

Reply:
I suppose that is a fair point. This either simplifies things or makes the same amount of work, and to your point it should be pretty much only us using it. "Us" being Fyne contributors. Or perhaps folks doing something very fancy, in which case "they" presumably know what they are doing.

I retract my opposition, so I would now say that I have no strong preference either way.

Reply:
If the desire is to have the complement of the `Register(scheme, URI)`, then `Lookup(scheme) URI` is surely a more appropriate name?
My understanding is that `ForURI` is a convenience function, in which case we pass the most convenient parameter.
I would prefer that we use one or the other (or both) rather than something in between.

I had a quick grep of the code and in around 90% if the calls we are working with a URI - so by moving to scheme as the lookup we are adding code to every one of those lines.
I don't see the compelling reason to be changing the API, particularly after many days of reviews.

Reply:
> If the desire is to have the complement of the Register(scheme, URI), then Lookup(scheme) URI is surely a more appropriate name?

Yes, assuming you meant s/URI/Repository, that was my original suggestion: https://github.com/fyne-io/fyne/pull/1768#issuecomment-758809328

I propose a compromise;
- Add `ForScheme(string) Repository` or `Lookup(string) Repository` which will return the Repository for the given scheme, and eliminates the allocation on the line this thread is attach to.
- Make `ForURI(URI) Repository` a helper function that is implemented simply as `return Lookup(uri.Scheme())` which the rest of the call sites can continue to use without modification.

Reply:
I think either @stuartmscott 's suggestion or just using `Lookup(string)` would both be fine.

Reply:
Apologies, yes I meant Repository in the realm of Register/Lookup.
The proposed approach sounds good, that's what I envisaged when I was saying "(or both)".

Reply:
I have implemented @stuartmscott 's suggestion in 184d27b8a5552fa0eaf258e4ab7bb35460ee8132.